includes/class-wc-coupon.php

```diff
@@ -54,97 +78,241 @@ class WC_Coupon {
 	const WC_COUPON_SUCCESS                          = 200;
 	const WC_COUPON_REMOVED                          = 201;
 
-	/** @public string Coupon code. */
-	public $code   = '';
+	/**
+	 * Coupon constructor. Loads coupon data.
+	 * @param  mixed $code code of the coupon to load
+	 */
+	public function __construct( $code = '' ) {
+		if ( $code instanceof WC_Coupon ) {
+			$this->read( absint( $code->get_id() ) );
+		} elseif ( $coupon = apply_filters( 'woocommerce_get_shop_coupon_data', false, $code ) ) {
```

Comment:
I hate this logic from long ago - I wonder if we can remove/deprecate it now since you could create a new coupon instance, set props, and just not save it.